cloud-controller-manager/do/resources.go

```diff
@@ -34,62 +35,211 @@ import (
-	gclient      *godo.Client
-	svcLister    v1lister.ServiceLister
+	clusterID string
+	kclient   kubernetes.Interface
+	gclient   *godo.Client
+	svcLister v1lister.ServiceLister
+
+	resources *resources
 }
 
-// NewResourcesController returns a new services controller responsible for
-// re-applying cluster IDs on DO load-balancers.
-func NewResourcesController(clusterIDTag string, inf v1informers.ServiceInformer,
-	k kubernetes.Interface, g *godo.Client) *ResourcesController {
+// NewResourcesController returns a new controller responsible for managing
+// DigitalOcean cloud resources.
+func NewResourcesController(
+	clusterID string,
+	r *resources,
+	inf v1informers.ServiceInformer,
+	k kubernetes.Interface,
+	g *godo.Client,
+) *ResourcesController {
 	return &ResourcesController{
-		clusterIDTag: clusterIDTag,
-		kclient:      k,
-		gclient:      g,
-		svcLister:    inf.Lister(),
+		clusterID: clusterID,
+		resources: r,
+		kclient:   k,
+		gclient:   g,
+		svcLister: inf.Lister(),
 	}
 }
 
 // Run starts the resources controller. It watches over DigitalOcean resources
 // making sure the right tags are set.
 func (r *ResourcesController) Run(stopCh <-chan struct{}) {
-	ticker := time.NewTicker(controllerSyncPeriod)
-	defer ticker.Stop()
-	// Do not wait for initial tick to pass; run immediately for reduced sync
-	// latency.
-	for tickerC := ticker.C; ; {
-		if err := r.sync(); err != nil {
-			glog.Errorf("failed to sync load-balancer tags: %s", err)
+	syncResourcesTicker := time.NewTicker(controllerSyncResourcesPeriod)
+	defer syncResourcesTicker.Stop()
+
+	syncTagsTicker := time.NewTicker(controllerSyncTagsPeriod)
+	defer syncTagsTicker.Stop()
+
+	go func() {
+		// Do not wait for initial tick to pass; run immediately for reduced sync
+		// latency.
+		for tickerC := syncResourcesTicker.C; ; {
+			if err := r.syncResources(); err != nil {
+				glog.Errorf("failed to sync cloud resources: %s", err)
+			}
+			select {
+			case <-tickerC:
+				continue
+			case <-stopCh:
+				return
+			}
 		}
-		select {
-		case <-tickerC:
-			continue
-		case <-stopCh:
-			return
+	}()
+
+	go func() {
+		// Do not wait for initial tick to pass; run immediately for reduced sync
+		// latency.
+		for tickerC := syncTagsTicker.C; ; {
+			if err := r.syncTags(); err != nil {
+				glog.Errorf("failed to sync load-balancer tags: %s", err)
+			}
+			select {
+			case <-tickerC:
+				continue
+			case <-stopCh:
+				return
+			}
 		}
-	}
+	}()
 }
 
-func (r *ResourcesController) sync() error {
-	// Fetch all load-balancers.
-	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+func (r *ResourcesController) syncResources() error {
+	ctx, cancel := context.WithTimeout(context.Background(), syncResourcesTimeout)
 	defer cancel()
-	lbs, _, err := r.gclient.LoadBalancers.List(ctx, &godo.ListOptions{})
+
+	glog.V(2).Info("syncing droplet resources.")
+	droplets, err := allDropletList(ctx, r.gclient)
+	if err != nil {
+		return err
```

Comment:
Can we change the logic so that failure to sync the droplets does not shortcut the method by returning an error?

IMHO, the method should not return an error but emit events when some of the resources cannot be sycned so that other resources still stand a chance to get synched.

Reply:
Yeah, switched to just logging for now. I can update to emit events in another PR when we want/need to act on these failures.